### ISSUE TITLE
Error: Cannot read property 'joinKey' of undefined

### ISSUE THREAD

**Issue by xilanhthanki** [ractivejs/ractive#3224]
I use Ractive with Backbone.js and I recently upgrade Ractive version from 0.7 to 0.9.13 
Before the upgrade, everything is fine.

Sometimes when I update the Backbone model, the error **Cannot read property 'joinKey' of undefined** occurs. This error in function ModelBase__proto__.joinAll

In some cases, I fixed by using .clone() method to the Backbone model. But in other cases, it has no effect. 

Do you please give me some suggestions to handle this issue?

Thanks.

**Comment by ghost** [comment]
Can we get a minimal reproduction of the issue. Also, are you using any adaptors? Just want to verify if issue is Ractive's, Backbone's or something else. I recently worked with the ractive-backbone adaptor with the latest Backbone and Ractive and have not seen anything like this so far.

**Comment by xilanhthanki** [comment]
I cannot public all code becauce the privacy of project.

The code cause error: 
```
return formItems.findWhere({
    itemNo: sort.get('itemNo'),
});
```
After this code, I change the property of the Backbone model. Then the error appears.

If I add clone() like this:

```
return formItems.findWhere({
    itemNo: sort.get('itemNo'),
}).clone();
```
Then the error gone.

Some case it works fine. But some other case it doesn't.
Don't understand the root cause of the error.

**Comment by ceremcem** [comment]
@xilanhthanki Your above description may or may not help to solve the problem, but I wanted to put a general recommendation, if you don't mind:

In my understanding, active Ractive users work on large projects and I do definitely know that sometimes it's a very time consuming task to create a minimal reproduction code in the [Playground](https://ractive.js.org/playground). But that's fair, because the investigator (@fskreuz, @evs-chris and the Ractive users) has to have something to work on/mull over. IMHO, a reproduction code is equal to a "pages of description of a problem". 

Bonus: If you actually try to make it a habit to create a reproduction code first, you will often find yourself pressing the "cancel button" while you are about to create a new issue. (guess how I might know :) )

**Comment by xilanhthanki** [comment]
Thanks @ceremcem for the recommendation. Highly appreciate it.
I cannot make the reproduction code in the Playground cause the code in the project too complex and the error may come from any corners.
Maybe you can close this issue. Thanks for your support.

**Comment by evs-chris** [comment]
Based on the stack trace, I'm not really sure _how_ this is happening. Looking through everywhere that calls `joinAll`, there should be no case where the model can not be an actual model as far as I can tell. It _would_ be tremendously helpful to be able to step through and see what's going wrong. If you can't reduce it to a smaller case or share code publicly, could you share it privately?

Failing that, can you tell me what `formItems` and `sort` are? I have no experience with backbone, but I may be able to stumble through it given some more details.

**Comment by xilanhthanki** [comment]
Unfortunately the project is huge because it have been run few years ago. So I don't know how to share with you. There're some problems about performance so we have to do many things to improve. My work is doing version up from 0.7 to 0.9.13 and fix all issues appears.

Btw the `formItems ` is a Backbone Collection and the `itemNo: sort.get('itemNo')` is just a condition to find a model in collection.
I found this in Backbone API: http://backbonejs.org/#Collection-findWhere

The function return a Backbone model (with some changed values) then the error occurs here. After cloning the model, the error disappears. So my guess is the mutation of the Backbone model causes this issue but I'm really not sure.

I know the hypothesis is too little for you to investigate. Thanks you all for your support.

**Comment by xilanhthanki** [comment]
Addition hypothesis. The problem may occur in Backbone.js adaptor for Ractive
See my attached images

The good case: no prefix @
![good](https://user-images.githubusercontent.com/21329596/38478050-8446eea4-3be0-11e8-91d5-e04351ff1212.JPG)

The bad case: has prefix @
![bad](https://user-images.githubusercontent.com/21329596/38478053-876d080c-3be0-11e8-8bfc-0c5b09fb3f5b.JPG)

All the time we see the prefix @, the error occurs.

The selected attribute trigger error on view:
![view](https://user-images.githubusercontent.com/21329596/38478594-f9db1b2e-3be3-11e8-9fd7-b3e1af132b0d.JPG)

Hope this help. Thanks.

**Comment by evs-chris** [comment]
Gotcha. This is an incompatibility between the backbone adaptor and ractive >= 0.8. What's happened is that computed contexts are no longer kept at the model level, due to potential overlap wherein one model gets overwritten by another that happens to have the same source path but is a completely different model. What needs to happen to fix this is the adaptor needs to get a context helper rather than a prefixer in order to be able to update relative keypaths. Additionally, setting computed contexts has been removed, also I believe from version 0.8.

As a workaround, you can avoid using a computed context (wherever the `getSelectedItem()` is being called in the template) along with the backbone adaptor. Depending on what you're doing that may not be feasible. I'm guessing you're using that in an iteration, and if so, you might try jumping to 0.10.0 and the experimental source mapped `{{#each}}` block. A reference expression (`some.path[variable]`) can also avoid computed contexts in 0.9, and as of 0.10 you can use an array of keys to get to any model if you know the path. You can safely have a computed member of a reference expression e.g. `~/[getPathTo(item.1)]`, which combined with 0.10's support for arrays in reference expressions looks like it may also get you where you need to be.

**Comment by xilanhthanki** [comment]
Your explanation is very reasonable. But unfortunately we cannot do another version up right now. 
Cause the system have run in 0.9.13 for a long time. So I cannot measure what will be affected after another version up.

Btw I have done upgrading 0.9.13 to 0.10.0, then see errors everywhere :( So I think I cannot do that anymore.

Do you have another suggestions for 0.9.13?
Thanks

**Comment by evs-chris** [comment]
Your errors with 0.10 would likely be resolved by setting `resolveInstanceMembers: true` if going from 0.7 to 0.9 worked and 0.10 didn't. There are very few breaking changes between 0.9 and 0.10, and that's the most visible one from older versions. Alternatively, prefixing your method calls in event listeners with `@.` would achieve the same effect in a safer way.

Other than avoiding computed contexts, there's not much I can think of that would work as a workaround in 0.9. I'll look at the backbone adaptor to see if I can come up with a reasonable fix there.

**Comment by xilanhthanki** [comment]
Thanks. Solved the errors by setting resolveInstanceMembers: true
Let me try your workaround.

**Comment by xilanhthanki** [comment]
After investigating, This is what I found

In parent component, I'm writing this
`<FormSelectbox options="{{form.items}}" selected="{{getFormItem(this)}}" index="{{index}}">`

Then in child component, everytime I use `selected` property, the error always occurs. Although I just use `console.log(selected)` only.

----------------

So instead of use `selected` property directly, I pass `itemNo` to child component, then do `findWhere` inside child component. The error doesn't appear anymore. This fix also works on 0.9
This may be inconvenience but I think I have to do this in my project.

Thank you all for your support.

**Comment by evs-chris** [comment]
Pushing this over to #2512.
